ibis/backends/flink/compiler.py

```diff
@@ -0,0 +1,642 @@
+"""Flink Ibis expression to SQL compiler."""
+
+from __future__ import annotations
+
+from functools import singledispatchmethod
+
+import sqlglot as sg
+import sqlglot.expressions as sge
+from sqlglot.dialects import Hive
+from sqlglot.dialects.dialect import rename_func
+
+import ibis.common.exceptions as com
+import ibis.expr.datatypes as dt
+import ibis.expr.operations as ops
+from ibis.backends.base.sqlglot.compiler import STAR, SQLGlotCompiler, paren
+from ibis.backends.base.sqlglot.datatypes import FlinkType
+from ibis.backends.base.sqlglot.rewrites import (
+    exclude_unsupported_window_frame_from_ops,
+    exclude_unsupported_window_frame_from_rank,
+    exclude_unsupported_window_frame_from_row_number,
+    rewrite_first_to_first_value,
+    rewrite_last_to_last_value,
+    rewrite_sample_as_filter,
+)
+
+
+class Flink(Hive):
+    class Generator(Hive.Generator):
+        pass
+
+    class Tokenizer(Hive.Tokenizer):
+        # In Flink, embedded single quotes are escaped like most other SQL
+        # dialects: doubling up the single quote
+        #
+        # We override it here because we inherit from Hive's dialect and Hive
+        # uses a backslash to escape single quotes
+        STRING_ESCAPES = ["'"]
+
+
+Flink.Generator.TYPE_MAPPING |= {
+    sge.DataType.Type.TIME: "TIME",
+}
+
+Flink.Generator.TRANSFORMS |= {
+    sge.Stddev: rename_func("stddev_samp"),
+    sge.StddevPop: rename_func("stddev_pop"),
+    sge.StddevSamp: rename_func("stddev_samp"),
+    sge.Variance: rename_func("var_samp"),
+    sge.VariancePop: rename_func("var_pop"),
+    sge.Array: (
+        lambda self,
+        e: f"ARRAY[{', '.join(arg.sql(self.dialect) for arg in e.expressions)}]"
+    ),
+    sge.ArrayConcat: rename_func("array_concat"),
+    sge.Length: rename_func("char_length"),
+    sge.TryCast: lambda self,
+    e: f"TRY_CAST({e.this.sql(self.dialect)} AS {e.to.sql(self.dialect)})",
+    sge.DayOfYear: rename_func("dayofyear"),
+    sge.DayOfWeek: rename_func("dayofweek"),
+    sge.DayOfMonth: rename_func("dayofmonth"),
+}
+
+
+class FlinkCompiler(SQLGlotCompiler):
+    quoted = True
+    dialect = Flink
+    type_mapper = FlinkType
+    rewrites = (
+        rewrite_sample_as_filter,
+        exclude_unsupported_window_frame_from_row_number,
+        exclude_unsupported_window_frame_from_ops,
+        exclude_unsupported_window_frame_from_rank,
+        rewrite_first_to_first_value,
+        rewrite_last_to_last_value,
+        *SQLGlotCompiler.rewrites,
+    )
+
+    @property
+    def NAN(self):
+        raise NotImplementedError("Flink does not support NaN")
+
+    @property
+    def POS_INF(self):
+        raise NotImplementedError("Flink does not support Infinity")
+
+    NEG_INF = POS_INF
+
+    def _aggregate(self, funcname: str, *args, where):
+        func = self.f[funcname]
+        if where is not None:
+            # FILTER (WHERE ) is broken for one or both of:
```

Comment:
We can apply rewrites for this case, though I guess this is not high priority.

Reply:
I think rewrites might be a bit overkill here when this implementation suffices 😅